### PR TITLE
[SPARK-34939][CORE] Throw fetch failure exception when unable to deserialize broadcasted map statuses

### PR DESCRIPTION
**Pull request by viirya**
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This patch catches `IOException`, which is possibly thrown due to unable to deserialize map statuses (e.g., broadcasted value is destroyed), when deserilizing map statuses. Once `IOException` is caught, `MetadataFetchFailedException` is thrown to let Spark handle it.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

One customer encountered application error. From the log, it is caused by accessing non-existing broadcasted value. The broadcasted value is map statuses. E.g.,

```
[info]   Cause: java.io.IOException: org.apache.spark.SparkException: Failed to get broadcast_0_piece0 of broadcast_0                                                                                                                  
[info]   at org.apache.spark.util.Utils$.tryOrIOException(Utils.scala:1410)                                                                                                                                                            
[info]   at org.apache.spark.broadcast.TorrentBroadcast.readBroadcastBlock(TorrentBroadcast.scala:226)                                                                                                                                 
[info]   at org.apache.spark.broadcast.TorrentBroadcast.getValue(TorrentBroadcast.scala:103)                                                                                                                                           
[info]   at org.apache.spark.broadcast.Broadcast.value(Broadcast.scala:70)                                                                                                                                                             
[info]   at org.apache.spark.MapOutputTracker$.$anonfun$deserializeMapStatuses$3(MapOutputTracker.scala:967)                                                                                                                           
[info]   at org.apache.spark.internal.Logging.logInfo(Logging.scala:57)                                                                                                                                                                
[info]   at org.apache.spark.internal.Logging.logInfo$(Logging.scala:56)                                                                                                                                                               
[info]   at org.apache.spark.MapOutputTracker$.logInfo(MapOutputTracker.scala:887)                                                                                                                                                     
[info]   at org.apache.spark.MapOutputTracker$.deserializeMapStatuses(MapOutputTracker.scala:967)                                                                                                                                                                                                                                                                                 
```


There is a race-condition. After map statuses are broadcasted and the executors obtain serialized broadcasted map statuses. If any fetch failure happens after, Spark scheduler invalidates cached map statuses and destroy broadcasted value of the map statuses. Then any executor trying to deserialize serialized broadcasted map statuses and access broadcasted value, `IOException` will be thrown. Currently we don't catch it in `MapOutputTrackerWorker` and above exception will fail the application.

Normally we should throw a fetch failure exception for such case. Spark scheduler will handle this.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Unit test.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 1]
```diff
@@ -100,7 +100,7 @@ private class ShuffleStatus(numPartitions: Int) extends Logging {
    * broadcast variable in order to keep it from being garbage collected and to allow for it to be
    * explicitly destroyed later on when the ShuffleMapStage is garbage-collected.
    */
-  private[this] var cachedSerializedBroadcast: Broadcast[Array[Byte]] = _
+  private[spark] var cachedSerializedBroadcast: Broadcast[Array[Byte]] = _
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:103]: Expose this for test.

**Review by HyukjinKwon — COMMENTED** [review]
Looks fine. cc @mridulm @Ngone51 FYI

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 2]
```diff
@@ -843,7 +843,13 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
         if (fetchedStatuses == null) {
           logInfo("Doing the fetch; tracker endpoint = " + trackerEndpoint)
           val fetchedBytes = askTracker[Array[Byte]](GetMapOutputStatuses(shuffleId))
-          fetchedStatuses = MapOutputTracker.deserializeMapStatuses(fetchedBytes, conf)
+          try {
+            fetchedStatuses = MapOutputTracker.deserializeMapStatuses(fetchedBytes, conf)
+          } catch {
+            case e: SparkException =>
```
Comment (dbtsai) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:849]: The failure could be `DIRECT`, how can you ensure it's only catching exception from broadcast?

Comment (dbtsai) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:849]: Oh, never mind. I saw the code in the following.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 3]
```diff
@@ -953,13 +959,19 @@ private[spark] object MapOutputTracker extends Logging {
...
-        logInfo("Broadcast mapstatuses size = " + bytes.length +
-          ", actual size = " + bcast.value.length)
-        // Important - ignore the DIRECT tag ! Start from offset 1
-        deserializeObject(bcast.value, 1, bcast.value.length - 1).asInstanceOf[Array[MapStatus]]
+        try {
+          // deserialize the Broadcast, pull .value array out of it, and then deserialize that
+          val bcast = deserializeObject(bytes, 1, bytes.length - 1).
+            asInstanceOf[Broadcast[Array[Byte]]]
+          logInfo("Broadcast mapstatuses size = " + bytes.length +
+            ", actual size = " + bcast.value.length)
```
Comment (dbtsai) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:967]: Maybe we should move line 964 to 967 out of the try block like in `DIRECT` case.

**Comment by Ngone51** [comment]
This seems to be the same issue with https://github.com/apache/spark/pull/27604. cc @liupc @cloud-fan

File: core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala [hunk 4]
```diff
@@ -68,5 +68,6 @@ private[spark] class FetchFailedException(
 private[spark] class MetadataFetchFailedException(
     shuffleId: Int,
     reduceId: Int,
-    message: String)
-  extends FetchFailedException(null, shuffleId, -1L, -1, reduceId, message)
+    message: String,
+    cause: Throwable = null)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala:72]: I don't see `cause` is used anywhere. Shall we covert the `cause` to stack string and append to the `message`?

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 3 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:967]: This is for the need of writing the test case. In the test case, if we call `getStatuses`, the mapoutput tracker worker will ask tracker master for new broadcasted value. So we cannot test the situation we need.

File: core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala (hunk 4 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala:72]: Sure, I just want to keep original stack trace.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 5]
```diff
@@ -953,13 +960,19 @@ private[spark] object MapOutputTracker extends Logging {
...
+          // Important - ignore the DIRECT tag ! Start from offset 1
+          deserializeObject(bcast.value, 1, bcast.value.length - 1).asInstanceOf[Array[MapStatus]]
+        } catch {
+          case e: IOException =>
+            logError("Exception encountered during deserializing broadcasted map statuses: ", e)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:973]: If this is recoverable, maybe shall we lower the level to `Warn`?

Comment (viirya) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:973]: Okay, makes sense.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by viirya** [comment]
@dbtsai @Ngone51 Any more comments? Otherwise I will merge this in days. I think there is conflict in 2.4, I will make a backport after.

**Comment by viirya** [comment]
Thanks @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Merged to master/3.1/3.0.

**Comment by viirya** [comment]
Thank all!

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 6]
```diff
@@ -953,13 +960,19 @@ private[spark] object MapOutputTracker extends Logging {
...
+          val bcast = deserializeObject(bytes, 1, bytes.length - 1).
+            asInstanceOf[Broadcast[Array[Byte]]]
+          logInfo("Broadcast mapstatuses size = " + bytes.length +
+            ", actual size = " + bcast.value.length)
+          // Important - ignore the DIRECT tag ! Start from offset 1
+          deserializeObject(bcast.value, 1, bcast.value.length - 1).asInstanceOf[Array[MapStatus]]
+        } catch {
+          case e: IOException =>
+            logWarning("Exception encountered during deserializing broadcasted map statuses: ", e)
+            throw new SparkException("Unable to deserialize broadcasted map statuses", e)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:974]: shall we throw MetadataFetchFailedException directly here?

Comment (viirya) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:974]: Throw `MetadataFetchFailedException` here and catch it and rethrow?

Comment (viirya) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:974]: Oh, recall why I did this. To construct `MetadataFetchFailedException` needs `shuffleId`.

I choose to not throw `MetadataFetchFailedException` as `deserializeMapStatuses ` doesn't have `shuffleId` and doesn't need it at all.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:974]: ah I see, thanks for the explanation!

**Comment by mridulm** [comment]
Got to this a bit late, looks good to me. Nice catch @viirya !

**Comment by viirya** [comment]
Thanks @cloud-fan @mridulm

**Comment by Ngone51** [comment]
Late LGTM.

**Pull request closed without merging**